packages/eslint-plugin/src/rules/no-base-to-string.ts

```diff
@@ -82,6 +89,50 @@ export default createRule<Options, MessageIds>({
       });
     }
 
+    function checkExpressionForArrayJoin(
+      node: TSESTree.Node,
+      type: ts.Type,
+    ): void {
+      const certainty = collectJoinCertainty(type);
+
+      if (certainty === Usefulness.Always) {
+        return;
+      }
+
+      context.report({
+        node,
+        messageId: 'baseArrayJoin',
+        data: {
+          name: context.sourceCode.getText(node),
+          certainty,
+        },
+      });
+    }
+
+    function collectJoinCertainty(type: ts.Type): Usefulness {
+      const typeParts = tsutils.typeParts(type);
+
+      const isArrayOrTuple = typeParts.every(
+        part => checker.isArrayType(part) || checker.isTupleType(part),
+      );
+      if (!isArrayOrTuple) {
+        return Usefulness.Always;
+      }
+
+      if (checker.isTupleType(type)) {
+        const typeArgs = checker.getTypeArguments(type);
+        const certainties = typeArgs.map(t => collectToStringCertainty(t));
+        if (certainties.some(certainty => certainty === Usefulness.Never)) {
+          return Usefulness.Never;
+        }
+      }
+      const elemType = type.getNumberIndexType();
```

Comment:
If you write this in a way such that `checker.isArrayType(type)` is manifestly true, then feel free to use a `nullThrows` here in order to satisfy coverage